activerecord/lib/active_record/insert_all.rb

```diff
@@ -0,0 +1,185 @@
+        primary_keys + model.readonly_attributes.to_a
+      end
+
+      def unique_by_columns
+        unique_by ? unique_by.fetch(:columns).map(&:to_s) : []
+      end
+
+      def primary_keys
+        Array.wrap(model.primary_key)
+      end
+
+      module SqlBuilders
+        class Generic
+          def self.build_sql(*args)
+            new(*args).to_sql
+          end
+
+          def initialize(model, keys, values_list, returning:, on_duplicate:, unique_by:, updatable_columns:)
+            @model, @connection, @keys, @values_list = model, model.connection, keys, values_list
+            @returning, @on_duplicate, @unique_by, @updatable_columns = returning, on_duplicate, unique_by, updatable_columns
+          end
+
+          attr_reader :model, :connection, :keys, :values_list,
+                      :returning, :unique_by, :updatable_columns
+
+          def skip_duplicates?
+            @on_duplicate == :skip
+          end
+
+          def update_duplicates?
+            @on_duplicate == :update
+          end
+
+          def to_sql
+            sql = [ "INSERT ", into_sql, values_sql ]
+
+            if skip_duplicates? || update_duplicates?
+              sql.push " ON CONFLICT", conflict_target_sql
+              sql.push " DO NOTHING" if skip_duplicates?
+              sql.push " DO UPDATE SET ", update_columns_sql if update_duplicates?
+            end
+
+            sql.push " RETURNING ", quote_columns(returning).join(",") if returning
+            sql.join
+          end
+
+          protected
+            def into_sql
+              [ "INTO ", model.quoted_table_name, "(", key_list_sql, ") " ].join
+            end
+
+            def values_sql
+              Arel::InsertManager.new.create_values_list(values_list).to_sql
+            end
+
+            def key_list_sql
+              quote_columns(keys).join(",")
+            end
+
+            def quote_columns(columns)
+              columns.map(&connection.method(:quote_column_name))
+            end
+
+            def conflict_target_sql
+              return unless conflict_columns
+              sql = [ "(", quote_columns(conflict_columns).join(","), ")" ]
+              sql.push " WHERE ", where if where
+              sql.join
+            end
+
+            def conflict_columns
+              @conflict_columns ||= begin
+                conflict_columns = unique_by.fetch(:columns) if unique_by
+                conflict_columns ||= Array.wrap(model.primary_key) if update_duplicates?
+                conflict_columns
+              end
+            end
+
+            def where
+              unique_by && unique_by[:where]
+            end
+
+            def update_columns_sql
+              quote_columns(updatable_columns).map { |column| "#{column} = excluded.#{column}" }.join(",")
+            end
+        end
+
+        class MySQL < Generic
+          def to_sql
+            sql = [ "INSERT " ]
+            sql.push "IGNORE " if skip_duplicates?
```

Comment:
`INSERT IGNORE` is unsafe operation, it ignores not only UNIQUE constraint but also all restrictions.

```mysql
root@localhost [test] > create table t (c varchar(5));
Query OK, 0 rows affected (0.03 sec)

root@localhost [test] > insert into t values ('abcdef');
ERROR 1406 (22001): Data too long for column 'c' at row 1

root@localhost [test] > insert ignore into t values ('abcdef');
Query OK, 1 row affected, 1 warning (0.00 sec)

Warning (Code 1265): Data truncated for column 'c' at row 1

root@localhost [test] > select * from t;
+-------+
| c     |
+-------+
| abcde |
+-------+
1 row in set (0.00 sec)
```

Reply:
Thank you so much for pointing that out, @kamipo!

[Looks like](https://stackoverflow.com/a/548570) it should be easy enough to solve for. We'll just use `ON DUPLICATE KEY UPDATE` but then just do a NULL update.

I'll push up a fix for that.